### PR TITLE
Shells in GitHub Actions

### LINKED ISSUE #10: The README instructions may contain wrong syntax
For example, the windows cmd instructions don't work. I tried in a github action the syntax was wrong.

### PR DESCRIPTION
**Pull request by dbeatty10**
Resolves #10

## Overview
For multiple environments, we are using GitHub Actions to verify that the provided commands for installing into a virtual environment work as expected.

We are testing the following combinations within GitHub Actions:
- ✅ `bash` for Linux and Mac containers
- ✅ Each of `powershell` and `cmd` using a Windows container for each set of commands
- ✅ `pwsh` to test PowerShell Core in a Linux container
- ❌ Leaving zsh, fish, csh, tcsh, and PowerShell Core as not covered by GitHub Actions (unless/until we discover an easy way to do so)

## Result
We are able to see the expected output for all environments except for Windows cmd.exe. For some reason, that environment doesn't display the log output. But it doesn't error, so maybe that means it worked 🤷 

## Manual steps
After validation via GitHub actions, the commands need to be copy-pasted into the `README.md` file.

## Background
Python provides [operating system and shell-specific instructions](https://docs.python.org/3/library/venv.html#creating-virtual-environments) for creating and activating virtual environments:

image

### CODE REVIEW

File: .github/workflows/validate_on_platforms.yml [hunk 1]
```diff
@@ -47,30 +40,47 @@ jobs:
...
             source venv/bin/activate
-        - name: Upgrade pip
-          run: |
             venv/bin/python3 -m pip install --upgrade pip
-        - name: Install dependencies
-          run: |
             python3 -m pip install -r requirements.txt
-        - name: Run all the dbt operations(seed, run, test)
-          run: |
+            source venv/bin/activate
```
Comment (sungchun12) [inline .github/workflows/validate_on_platforms.yml:49]:
```suggestion

```

Comment (sungchun12) [reply .github/workflows/validate_on_platforms.yml:49]: You don't need to activate the virtual environment again. I proved this already in a previous PR. 
https://github.com/dbt-labs/jaffle_shop_duckdb/runs/7137404465?check_suite_focus=true

**Review by sungchun12 — CHANGES_REQUESTED** [review]
Can you run the github actions with the committed suggestions?

**Review by sungchun12 — CHANGES_REQUESTED** [review]
batched these suggestions together

File: .github/workflows/validate_on_platforms.yml [hunk 2]
```diff
@@ -8,35 +8,28 @@ on:
...
-          duckcli jaffle_shop.duckdb -e "select * from customers where customer_id = 42"
+      name: Validate Getting Started works on macOS
+      runs-on: macos-latest
+      steps:
+        - uses: actions/checkout@v2
+        - name: Set up Python
+          uses: actions/setup-python@v1
+          with:
+            python-version: "3.8.x"
+        - name: Install dbt and DuckDB in a virtual environment
```
Comment (sungchun12) [inline .github/workflows/validate_on_platforms.yml]:
```suggestion
        - name: Full Deploy Commands
```

File: .github/workflows/validate_on_platforms.yml [hunk 3]
```diff
@@ -47,30 +40,47 @@ jobs:
           uses: actions/setup-python@v1
           with:
             python-version: "3.8.x"
-        - name: Setup Virtual Environment
+        - name: Install dbt and DuckDB in a virtual environment
```
Comment (sungchun12) [inline .github/workflows/validate_on_platforms.yml]:
```suggestion
        - name: Full Deploy Commands
```

File: .github/workflows/validate_on_platforms.yml [hunk 4]
```diff
@@ -47,30 +40,47 @@ jobs:
...
+  build-linux-powershell-core:
+      name: Validate Getting Started works on Linux PowerShell Core
+      runs-on: ubuntu-latest
+      steps:
+        - uses: actions/checkout@v2
+        - name: Set up Python
+          uses: actions/setup-python@v1
+          with:
+            python-version: "3.8.x"
+        - name: Install dbt and DuckDB in a virtual environment
```
Comment (sungchun12) [inline .github/workflows/validate_on_platforms.yml]:
```suggestion
        - name: Full Deploy Commands
```

File: .github/workflows/validate_on_platforms.yml [hunk 5]
```diff
@@ -81,21 +91,38 @@ jobs:
         - name: Setup Virtual Environment
```
Comment (sungchun12) [inline .github/workflows/validate_on_platforms.yml]:
```suggestion
        - name: Full Deploy Commands
```

File: .github/workflows/validate_on_platforms.yml [hunk 6]
```diff
@@ -81,21 +91,38 @@ jobs:
...
+  build-windows-cmd:
+      name: Validate Getting Started works on Windows cmd.exe
+      runs-on: windows-latest
+      steps:
+        - uses: actions/checkout@v2
+        - name: Set up Python
+          uses: actions/setup-python@v1
+          with:
+            python-version: "3.8.x"
+        - name: Setup Virtual Environment
```
Comment (sungchun12) [inline .github/workflows/validate_on_platforms.yml]:
```suggestion
        - name: Full Deploy Commands
```

**Review by sungchun12 — CHANGES_REQUESTED** [review]
Rename the job step to be clearer in what the steps are achieving

File: .github/workflows/validate_on_platforms.yml (hunk 1 above)
Comment (dbeatty10) [reply .github/workflows/validate_on_platforms.yml:49]: In some cases (like a fresh GitHub Actions container), a reactivation may not be needed. But it is needed in certain circumstances, namely certain local environments (like mine). The reactivation is a super fast command, and in the worst case it is a no-op, so there's no real downside to including it. To make the commands in GitHub Action match the README precisely, I put this line back in.

###  Circumstance in which it is necessary
I needed to reactivate the virtual environment on my machine. I have dbt-core installed in my environment via pipx. When creating a virtual environment, sourcing it, and doing an isolated install of dbt-core, I need to reactivate the environment in order to change where the shell is looking for `dbt`.

### Replicating within GitHub Actions
You might be able replicate this behavior in GitHub Actions by doing the following in a Linux container:
1. [Install pipx](https://pypa.github.io/pipx/installation/)
2. Install dbt Core (`python3 -m pip install dbt-core`)
3. List where it is installed (`which dbt`)
4. Do all the virtual environment steps and do another `which dbt`

Comment (sungchun12) [reply .github/workflows/validate_on_platforms.yml:49]: @dbeatty10 Can you reply with a screenshot of what happens on your local machine when you do NOT reactivate the virtual environment? I believe you and want internet strangers to understand why we're adding this command. I've personally never ran into the issue you illustrated, so I'm curious what it looks like on your machine.

Can you also add to the README on why the steps require reactivating the virtual environment is necessary? You can tweak the language in your `Circumstance in which it is necessary` section and place it at the bottom of the README somewhere.

**Comment by dbeatty10** [comment]
> Rename the job step to be clearer in what the steps are achieving

Good call for renaming each of these job steps so they are clear and match across environments. I made a commit that the contains those specific suggestions.

> Can you run the github actions with the committed suggestions?

We can't pull out the 2nd activation or it won't be guaranteed to work in all cases. See explanation here: https://github.com/dbt-labs/jaffle_shop_duckdb/pull/14#discussion_r913300981

So I'd like to move forward with the 2nd activation included so we can be confident that the instructions should work for all scenarios.

I'll request a re-review.

### Example needing the 2nd activation

I use `zsh` as my shell on my MacBook Pro, and I already had an alpha version of dbt Core 1.2 installed via [pipx](https://pypa.github.io/pipx/installation/):
```shell
$ which dbt
/Users/dbeatty/.pyenv/shims/dbt
```
```shell
$ dbt --version
Core:
  - installed: 1.2.0-a1
  - latest:    1.1.1    - Ahead of latest version!

Plugins:
  - bigquery:  1.2.0a1 - Ahead of latest version!
  - snowflake: 1.2.0a1 - Ahead of latest version!
  - redshift:  1.2.0a1 - Ahead of latest version!
  - postgres:  1.2.0a1 - Ahead of latest version!
```

Then I ran all the steps to create a virtual environment and install the requirements of our DuckDB-based Jaffle Shop repo:
```shell
$ python3 -m venv venv
$ source venv/bin/activate
(venv) $ venv/bin/python3 -m pip install --upgrade pip
(venv) $ python3 -m pip install -r requirements.txt
```

Let's examine where `dbt` is installed and which version it is reporting:
```shell
(venv) $ which dbt
/Users/dbeatty/projects/jaffle_duck/venv/bin/dbt
```

```shell
(venv) $ dbt --version
Core:
  - installed: 1.2.0-a1
  - latest:    1.1.1    - Ahead of latest version!

Plugins:
  - bigquery:  1.2.0a1 - Ahead of latest version!
  - snowflake: 1.2.0a1 - Ahead of latest version!
  - redshift:  1.2.0a1 - Ahead of latest version!
  - postgres:  1.2.0a1 - Ahead of latest version!
```

❌ That isn't what we expected -- something isn't right. 😢

So let's reactivate the virtual environment and try again...
```shell
(venv) $ source venv/bin/activate
```

```shell
(venv) $ dbt --version
Core:
  - installed: 1.1.1
  - latest:    1.1.1 - Up to date!

Plugins:
  - postgres: 1.1.1 - Up to date!
  - duckdb:   1.1.3 - Up to date!
```

✅ This is what we want -- the 2nd reactivation worked. 😎

**Comment by sungchun12** [comment]
@dbeatty10 the above looks great! Honestly, if you include that in a toggle at the bottom, that should suffice to explain to new readers WHY the second virtual environment activation is required. After you do that, I'll approve and squash and merge!

File: .github/workflows/validate_on_platforms.yml (hunk 1 above)
Comment (dbeatty10) [reply .github/workflows/validate_on_platforms.yml:49]: I don't really understand all the details of the shenanigans that `pyenv`, `pipx`, and Python `venv` do within the shell, so I don't actually know _why_ it is necessary 🤷 

The screenshots can be actually a little hard to read, so I already put the commands and output [here](https://github.com/dbt-labs/jaffle_shop_duckdb/pull/14#issuecomment-1174480708).

But here's some actual screenshots showing the relevant sections:

Error before re-activation:

image


Before vs. after re-activation:

image

**Comment by dbeatty10** [comment]
@sungchun12 I added that toggle at the bottom, so it's all ready for you 👍

**Review by sungchun12 — APPROVED** [review]

**Comment by sungchun12** [comment]
Beautiful job @dbeatty10 :)

**Pull request merged**